File: backend/app/models/offer.py
from sqlalchemy import Column, Integer, String, Text, Float, DateTime, ForeignKey, func
from sqlalchemy.orm import relationship

from ..database import Base


class Offer(Base):
    __tablename__ = "offers"

    id = Column(Integer, primary_key=True, index=True)
    tender_call_id = Column(Integer, ForeignKey("tender_calls.id"), nullable=False)
    company_id = Column(Integer, ForeignKey("companies.id"), nullable=False)
    submitted_by_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    montant = Column(Float, nullable=False)
    devise = Column(String(3), nullable=False, default="USD")
    delai_execution = Column(String(100), nullable=True)
    proposition_technique = Column(Text, nullable=True)
    proposition_financiere = Column(Text, nullable=True)
    statut = Column(String(50), nullable=False, default="submitted")
    score_technique = Column(Float, nullable=True)
    score_financier = Column(Float, nullable=True)
    score_total = Column(Float, nullable=True)
    date_soumission = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    tender_call = relationship("TenderCall", back_populates="offres")
    company = relationship("Company", back_populates="offres")
    submitted_by = relationship("User", back_populates="submitted_offers")
    documents = relationship("OfferDocument", back_populates="offer")
    evaluations = relationship("OfferEvaluation", back_populates="offer")
    marche_public = relationship("PublicContract", back_populates="offer", uselist=False)

File: frontend/src/types/offer.ts
import type { Currency } from "./currency";

export type OfferStatus = "draft" | "submitted" | "under_review" | "accepted" | "rejected" | "awarded";

export interface Offer {
  id: number;
  tender_call_id: number;
  company_id: number;
  submitted_by_id: number;
  montant: number;
  devise: Currency;
  delai_execution?: string | null;
  proposition_technique?: string | null;
  proposition_financiere?: string | null;
  statut: OfferStatus;
  score_technique?: number | null;
  score_financier?: number | null;
  score_commission?: number | null;
  score_total?: number | null;
  date_soumission: string;
  created_at: string;
  updated_at: string;
}

export interface OfferCreate {
  tender_call_id: number;
  company_id: number;
  montant: number;
  devise: Currency;
  delai_execution?: string;
  proposition_technique?: string;
  proposition_financiere?: string;
}

export type OfferUpdate = Partial<OfferCreate> & {
  statut?: OfferStatus;
  score_technique?: number;
  score_financier?: number;
  score_commission?: number;
  score_total?: number;
};
